math/src/main/scala/breeze/linalg/support/CanTraverseValues.scala

```diff
@@ -52,81 +65,105 @@ trait CanTraverseValues[From, A] {
 
 object CanTraverseValues {
 
-  trait ValuesVisitor[@specialized A] {
-    def visit(a: A)
-    def visitArray(arr: Array[A]):Unit = visitArray(arr, 0, arr.length, 1)
+  /**
+    * This trait can be implemented and used to apply a function over a breeze collection/tensor,
+    * using [[CanTraverseValues.traverse()]].
+    *
+    * @tparam V
+    */
+  trait ValuesVisitor[@specialized V] {
 
-    def visitArray(arr: Array[A], offset: Int, length: Int, stride: Int):Unit = {
+    def visit(a: V)
+
+    /** Visits an array assuming offset of zero and stride of 1 */
+    final def visitArray(arr: Array[V]): Unit = visitArray(arr, 0, arr.length, 1)
+
+    def visitArray(arr: Array[V], offset: Int, length: Int, stride: Int): Unit = {
       import spire.syntax.cfor._
       // Standard array bounds check stuff
       if (stride == 1) {
-        cforRange(offset until length + offset) { i =>
-          visit(arr(i))
-        }
+        cforRange(offset until length + offset) { i => visit(arr(i)) }
       } else {
-        cforRange(0 until length) { i =>
-          visit(arr(i * stride + offset))
-        }
+        cforRange(0 until length) { i => visit(arr(i * stride + offset)) }
       }
     }
-    def zeros(numZero: Int, zeroValue: A)
+
+    def zeros(numZero: Int, zeroValue: V): Unit
+
```

Comment:
visitZeros sounds great

On Fri, Apr 8, 2016 at 10:24 AM, Kenta Takagaki <notifications@github.com>
wrote:

> In math/src/main/scala/breeze/linalg/support/CanTraverseValues.scala
> <https://github.com/ktakagaki/breeze/pull/17#discussion_r59059931>:
>
> >        }
> >      }
> > -    def zeros(numZero: Int, zeroValue: A)
> > +
> > +    def zeros(numZero: Int, zeroValue: V): Unit
> > +
>
> I see.
>
> Are you happy with the zeros naming? It sounds a bit to me like a dense
> array constructor filled with zeros? How about visitZeros?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ktakagaki/breeze/pull/17/files/0ef0bf906e05475a253f9ef782e4f603688f9063#r59059931>
>